### PR TITLE
`torch.compile` and `eager` benchmarks for `softmax`

### PR DESCRIPTION
**Pull request by Priya2698**
Adds `torch.compile` and `eager` baseline benchmarks to be used in weekly benchmark runs.
Issue #1668.

Update:
Building off of @jacobhinkle's suggestion:
1. `torch.compile` and `eager` mode are combined in a single benchmark function using parametrization.
2. By default, only nvfuser benchmarks are run. To enable eager/compile, the flags `--benchmark-eager` and `benchmark-torchcompile` are used. 

I'll update the wiki to describe and detail benchmarking options.

Future changes:
In a separate PR, I plan on renaming `disable-benchmarking` to `disable-nvf-benchmarking` since we are adding more executors.

### CODE REVIEW

File: python_benchmarks/test_softmax_fwd.py [hunk 1]
```diff
@@ -63,8 +67,36 @@ def test_softmax_fwd_benchmark(
...
+@pytest.mark.parametrize("reduction_axis", [0, 1])
+def test_softmax_fwd_compile_benchmark(
+    benchmark,
+    size: tuple,
+    dtype: torch.dtype,
+    reduction_axis: int,
+):
+    clear_cuda_cache()
+    input = torch.randn(*size, device="cuda", dtype=dtype)
+    run_benchmark(benchmark, torch.compile(softmax_fwd_fn), [input, reduction_axis])
```
Comment (jacobhinkle) [inline python_benchmarks/test_softmax_fwd.py:102]: Feel free to ignore, but since these tests are nearly identical is it practical to parametrize a single test here over whether `torch.compile` is used or not? I am thinking something like
```python
@pytest.mark.parametrize("size", generate_input_sizes(dims=2))
@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
@pytest.mark.parametrize("reduction_axis", [0, 1])
@pytest.mark.parametrize("compile", [False, True], ids=["eager", "compile"])
def test_softmax_fwd_compile_benchmark(
    benchmark,
    size: tuple,
    dtype: torch.dtype,
    reduction_axis: int,
    compile: bool,
):
    clear_cuda_cache()
    input = torch.randn(*size, device="cuda", dtype=dtype)
    run_benchmark(
        benchmark,
        torch.compile(softmax_fwd_fn) if compile else softmax_fwd_fn,
        [input, reduction_axis])
```

Comment (jacobhinkle) [reply python_benchmarks/test_softmax_fwd.py:102]: Same for bwd tests above.

**Review by jacobhinkle — APPROVED** [review]
LGTM, just minor comments.

File: python_benchmarks/test_softmax_fwd.py (hunk 1 above)
Comment (Priya2698) [reply python_benchmarks/test_softmax_fwd.py:102]: I considered this. The main reason I split them is to be able to run them independently using filtering flags (`-k eager, compile`). I am looking into if I can filter on the parametrization flag in a clean way.
The other option could be to merge and use flags similar to `disable-validation` and `disable-benchmarking.`

File: python_benchmarks/test_softmax_bwd.py [hunk 2]
```diff
@@ -54,10 +54,15 @@ def softmax_bwd_fusion(
     fd.add_output(T19)
 
 
+def softmax_bwd_fn(inputs: list):  # [in_tensor, output, grads]
+    inputs[1].backward(inputs[2], retain_graph=True)
+    return inputs[0].grad
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply python_benchmarks/test_softmax_bwd.py:59]: Good point, will rename, thanks!

File: python_benchmarks/test_softmax_fwd.py (hunk 1 above)
Comment (jacobhinkle) [reply python_benchmarks/test_softmax_fwd.py:102]: You can use `-k` with the `ids` that print I believe. In this case the test prints like `test_softmax_fwd_compile_benchmark[eager-0-torch.float32-(16, 768)]`. I guess you'd want to rename the function to `test_softmax_fwd_benchmark` instead, but point being that the `ids` argument to `parametrize` helps to make these filterable.

Comment (jacobhinkle) [reply python_benchmarks/test_softmax_fwd.py:102]: On that topic, I actually had some trouble filtering on shapes because  of how they print with commas. For example try filtering down to only tests that run with size `(16, 6400)`. I still don't know how to use `-k` to do that without modifying the ids. In cases like that, you can use something like this though
```python
@pytest.mark.parametrize("size", generate_input_sizes(dims=2), ids=lambda s: '_'.join(str(si) for si in s))
def test_foo(
    ...
```
This will print names like `python_benchmarks/test_softmax_fwd.py::test_foo[16_6400]` which you can filter to using `pytest foo.py -k "foo and 16_6400]"`.

Comment (Priya2698) [reply python_benchmarks/test_softmax_fwd.py:102]: It is filterable on ids. But I am having to filter the entire test ID: `-k test_softmax_fwd_compile_benchmark[eager-0-torch.float32-(16, 768)]` instead of something like `-k eager`. Although, this can work too (`test_softmax_fwd_compile_benchmark[*eager*]`), my only reluctance is that it is quite verbose and test name dependent. I believe I can simplify it by adding parser options.

Comment (Priya2698) [reply python_benchmarks/test_softmax_fwd.py:102]: > On that topic, I actually had some trouble filtering on shapes because of how they print with commas. For example try filtering down to only tests that run with size `(16, 6400)`. I still don't know how to use `-k` to do that without modifying the ids. In cases like that, you can use something like this though
> 
> ```python
> @pytest.mark.parametrize("size", generate_input_sizes(dims=2), ids=lambda s: '_'.join(str(si) for si in s))
> def test_foo(
>     ...
> ```
> 
> This will print names like `python_benchmarks/test_softmax_fwd.py::test_foo[16_6400]` which you can filter to using `pytest foo.py -k "foo and 16_6400]"`.

That's a good point, let me see if I can change this without needing significant changes in the CI. @xwang233 has scripts to process the outputs generated by pytest and assumes the sizes to be integers at the moment.

File: python_benchmarks/conftest.py [hunk 3]
```diff
@@ -33,3 +46,27 @@ def pytest_make_parametrize_id(val):
...
+                item.add_marker(skip_eager)
+
+    if not run_torchcompile:
+        skip_torchcompile = pytest.mark.skip(
+            reason="need --benchmark-torchcompile option to run"
+        )
+        for item in items:
+            # If the benchmark has compile=True parameter (torch.compile mode), skip it.
+            if "compile" in item.callspec.params and item.callspec.params["compile"]:
+                item.add_marker(skip_torchcompile)
```
Comment (jacobhinkle) [inline python_benchmarks/conftest.py:72]: So this will skip any tests with a `compile=True` or `compile=False` option. Is our convention then that any eager mode benchmarks should have a `compile=False` option? We should document those conventions in the docstring of this function or maybe in the docstring of this module.

File: python_benchmarks/test_softmax_fwd.py [hunk 4]
```diff
@@ -63,8 +67,28 @@ def test_softmax_fwd_benchmark(
...
+@pytest.mark.parametrize("reduction_axis", [0, 1])
+def test_softmax_fwd_baseline_benchmark(
+    benchmark,
+    size: tuple,
+    dtype: torch.dtype,
+    reduction_axis: int,
+    compile: bool,
+):
+    clear_cuda_cache()
+    input = torch.randn(*size, device="cuda", dtype=dtype)
```
Comment (jacobhinkle) [inline python_benchmarks/test_softmax_fwd.py:89]: I think `torch.randn` can accept a tuple of ints for the shape parameter, so no asterisk is needed here.

File: python_benchmarks/test_softmax_bwd.py [hunk 5]
```diff
@@ -82,3 +87,25 @@ def test_softmax_bwd_benchmark(
...
+def test_softmax_bwd_baseline_benchmark(
+    benchmark,
+    size: tuple,
+    dtype: torch.dtype,
+    reduction_axis: int,
+    compile: bool,
+):
+    clear_cuda_cache()
+    input = torch.randn(*size, device="cuda", dtype=dtype, requires_grad=True)
+    grads = torch.randn(*size, device="cuda", dtype=dtype)
```
Comment (jacobhinkle) [inline python_benchmarks/test_softmax_bwd.py:105]: See comment below about asterisks.

File: python_benchmarks/conftest.py [hunk 6]
```diff
@@ -15,6 +15,19 @@ def pytest_addoption(parser):
         default=False,
         help="Disable benchmarking.",
     )
+    parser.addoption(
+        "--benchmark-eager",
+        action="store_true",
+        default=False,
```
Comment (jacobhinkle) [inline python_benchmarks/conftest.py:21]: Minor: I think `default=False` is not needed if `action="store_true"` is given.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Just a few suggestions.
